### PR TITLE
Ensure HttpClientConfig#connectionProvider returns the original provider used to create HttpClient

### LINKED ISSUE #2462: Use HttpConnectionProvider mutate method
## Motivation

I want to make new `ConnectionProvider` using `defaultHttpClient.configuration().connectionProvider().mutate()` operation.
However I can't use `mutate()` method.
```java
  return new ReactorClientHttpConnector(reactorResourceFactory, defaultHttpClient -> {
    // builder value is always null.
    Builder builder = defaultHttpClient.configuration().connectionProvider().mutate();

  }
}
```

Because HttpConnectionProvider(defaultHttpClient type is HttpConnectionProvider) don't override `mutate()` method and default `mutate()` method return null.
```java
@FunctionalInterface
public interface ConnectionProvider extends Disposable {
	@Nullable
	default Builder mutate() {
		return null;
	}
}
```

## Desired solution
```java
private ReactorClientHttpConnector customize(maxIdleTimeoutMillis, maxLifeTimeoutMillis) {
  return new ReactorClientHttpConnector(reactorResourceFactory, defaultHttpClient -> {
    ConnectionProvider provider = defaultHttpClient.configuration().connectionProvider()
        .mutate()
        .name("provider")
        .maxIdleTime(Duration.ofMillis(maxIdleTimeoutMillis))
        .maxLifeTime(Duration.ofMillis(maxLifeTimeoutMillis))
        .build();

    HttpClient httpClient = HttpClient.create(provider)
  }
}
```
I made PR https://github.com/reactor/reactor-netty/pull/2463

## Considered alternatives
I have a question. [Violetagg told me](https://gitter.im/reactor/reactor-netty?at=6311c13c999499629345befb) that `defaultHttpClient.configuration().connectionProvider()` should return the original ConnectionProvider.

So I made below test code. I think that `configuredConnectionProvider.http1ConnectionProvider`  return the original ConnectionProvider. Isn't it? Please tell me if I'm mistaken.
```java
@Test
void configuredConnectionProviderShouldReturnTheOriginalConnectionProvider() {
	ConnectionProvider provider = ConnectionProvider.create("provider");
	try {
		HttpClient client = HttpClient.create(provider);
		HttpConnectionProvider configuredConnectionProvider =
				(HttpConnectionProvider) client.configuration().connectionProvider();
		assertThat(provider.getClass().getName())
				.isEqualTo(configuredConnectionProvider.http1ConnectionProvider.getClass().getName());
	}
	finally {
		provider.disposeLater()
				.block(Duration.ofSeconds(5));
	}
}
```
스크린샷 2022-09-04 오후 12 56 37

### PR DESCRIPTION
**Pull request by violetagg**
Fixes #2462

### CODE REVIEW

**Review by pderop — APPROVED** [review]
LGTM

**Comment by yangbongsoo** [comment]
@violetagg I have a few questions. 
Before that, I'm not raising a question about your way.
I think it could be a better way because you know more about `reactor-netty` than I do.

I chose the method of minimizing the change And There are two things in HttpConnectionProvider, http1 and http2, so I thought it was right to consider both(https://github.com/reactor/reactor-netty/pull/2468). Isn't this method rather simple? Let me know if there's anything I didn't think of. I want to learn.

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java [hunk 1]
```diff
@@ -265,7 +264,7 @@ public void subscribe(CoreSubscriber<? super Connection> actual) {
 
 				AddressResolverGroup<?> resolver = _config.resolverInternal();
 
-				_config.connectionProvider()
+				_config.httpConnectionProvider()
```
Comment (yangbongsoo) [inline reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java:267]: I don't know if it's best to use the implementation directly without using the interface.

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConfig.java [hunk 2]
```diff
@@ -115,6 +115,11 @@ public ChannelOperations.OnSetup channelOperationsProvider() {
 		return (ch, c, msg) -> new HttpClientOperations(ch, c, cookieEncoder, cookieDecoder);
 	}
 
+	@Override
+	public ConnectionProvider connectionProvider() {
+		return httpConnectionProvider().http1ConnectionProvider();
```
[the remark this answers is not in the document]
Comment (violetagg) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConfig.java:120]: No, `Http2ConnectionProvider` is created based on the `ConnectionProvider` that is provided when you create the `HttpClient`.

See this

https://github.com/reactor/reactor-netty/blob/472c2a1d4283f9dda3a02feca2dd2d349a3b5eba/reactor-netty-http/src/main/java/reactor/netty/http/client/HttpConnectionProvider.java#L89

and this

https://github.com/reactor/reactor-netty/blob/472c2a1d4283f9dda3a02feca2dd2d349a3b5eba/reactor-netty-http/src/main/java/reactor/netty/http/client/Http2ConnectionProvider.java#L83

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConfig.java [hunk 3]
```diff
@@ -321,7 +326,7 @@ public WebsocketClientSpec websocketClientSpec() {
 	Function<String, String> uriTagValue;
 	WebsocketClientSpec websocketClientSpec;
 
-	HttpClientConfig(ConnectionProvider connectionProvider, Map<ChannelOption<?>, ?> options,
+	HttpClientConfig(HttpConnectionProvider connectionProvider, Map<ChannelOption<?>, ?> options,
```
[the remark this answers is not in the document]
Comment (violetagg) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConfig.java:329]: It is OK to work with more specific type when needed.

[the remark this answers is not in the document]
Comment (violetagg) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConfig.java:329]: For example:

The copy constructor for `ClientTransportConfig` is

https://github.com/reactor/reactor-netty/blob/472c2a1d4283f9dda3a02feca2dd2d349a3b5eba/reactor-netty-core/src/main/java/reactor/netty/transport/ClientTransportConfig.java#L170

However in `HttpClientConfig` we need to copy exactly `HttpClientConfig`, so we have:

https://github.com/reactor/reactor-netty/blob/472c2a1d4283f9dda3a02feca2dd2d349a3b5eba/reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConfig.java#L338

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java (hunk 1 above)
Comment (violetagg) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java:267]: It is ok, one and the same implementation is used before and now.

**Review by yangbongsoo — APPROVED** [review]

**Comment by violetagg** [comment]
@pderop @yangbongsoo Thanks for the review!

**Pull request merged**